service/history/replication/batchable_task.go

```diff
@@ -0,0 +1,157 @@
+package replication
+
+import (
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	ctasks "go.temporal.io/server/common/tasks"
+)
+
+type (
+	BatchableTask interface {
+		TrackableExecutableTask
+		// BatchWith task and return a new BatchableTask
+		BatchWith(task BatchableTask) (BatchableTask, error)
+		CanBatch() bool
+		// MarkUnbatchable will mark current task not batchable, so CanBatch() will return false
+		MarkUnbatchable()
+	}
+
+	batchedTask struct {
+		batchedTask     TrackableExecutableTask
+		individualTasks []TrackableExecutableTask
+		lock            sync.Mutex
+		state           batchState
+		// individualTaskHandler will be called when batched task was Nack, Reschedule, MarkPoisonPill
+		individualTaskHandler func(task TrackableExecutableTask)
+		logger                log.Logger
+		metricsHandler        metrics.Handler
+	}
+
+	batchState int
+)
+
+const (
+	batchStateOpen  = 0
+	batchStateClose = 1
+)
+
+var _ TrackableExecutableTask = (*batchedTask)(nil)
+
+func (w *batchedTask) QueueID() interface{} {
+	return w.batchedTask.QueueID()
+}
+
+func (w *batchedTask) TaskID() int64 {
+	return w.individualTasks[0].TaskID()
+}
+
+func (w *batchedTask) TaskCreationTime() time.Time {
+	return w.individualTasks[0].TaskCreationTime()
+}
+
+func (w *batchedTask) MarkPoisonPill() error {
+	if len(w.individualTasks) == 1 {
```

Comment:
why length matters here?

Reply:
When len is 1 means no other tasks are been batched into this task(the batchedTask is actually the single task itself), so we can mark it as poisonPill. When len is great than 1, means multiple tasks are batched together, in this case, we would like to give the individual tasks another try before mark it as poison pill. Same logic applies to Nack.

Reply:
I see. Do you also need a lock here? Could it add new task after the length check?

Reply:
From the lifecycle of the batchedTask, we should be okay without lock here. `MarkPoisonPill()` should be called after Execute() method. Once Execute() is called, the batchedTask status becomes `close` and it will not accepting any new task to be batched into current task.

Reply:
I see. Seems like it has some assumption on the lifecycle. Maybe check the task status to be close. But it depends on you.